Dart/src/com/jetbrains/lang/dart/ide/formatter/DartBlock.java

```diff
@@ -30,6 +35,8 @@
   private static final TokenSet LAST_TOKENS_IN_SWITCH_CASE = TokenSet.create(
     BREAK_STATEMENT, CONTINUE_STATEMENT, RETURN_STATEMENT
   );
+  private static final Key<DartBlock> DART_RECEIVER_KEY = Key.create("DART_RECEIVER_KEY");
```

Comment:
"user data" is a universal mechanism that is used when there are no other means to accomplish the task. User data driven logic is harder to read and understand, and also it is usually error-prone.